// sigaa-api/dist/bonds/sigaa-bond-factory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaBondFactory = void 0;

require("source-map-support/register");

var _sigaaStudentBond = require("./sigaa-student-bond");

var _sigaaTeacherBond = require("./sigaa-teacher-bond");

/**
 * Class to implements BondFactory.
 *
 * Serves to create bond instances.
 * @category Internal
 */
class SigaaBondFactory {
  constructor(httpFactory, parser, courseFactory, activityFactory) {
    this.httpFactory = httpFactory;
    this.parser = parser;
    this.courseFactory = courseFactory;
    this.activityFactory = activityFactory;
  }
  /**
   * Creates a student program instance.
   *
   * @param registration It is the student registration code
   * @param program It's the name of the student program, in Portuguese it is called "curso".
   * @param bondSwitchUrl If the user has more than one bond, the bond link will be used to change the bond
   */


  createStudentBond(registration, program, bondSwitchUrl) {
    let http;

    if (bondSwitchUrl) {
      http = this.httpFactory.createHttpWithBond(bondSwitchUrl);
    } else {
      http = this.httpFactory.createHttp();
    }

    return new _sigaaStudentBond.SigaaStudentBond(http, this.parser, this.courseFactory, this.activityFactory, program, registration, bondSwitchUrl);
  }
  /**
   * Creates a teacher bond instance.
   */


  createTeacherBond() {
    return new _sigaaTeacherBond.SigaaTeacherBond();
  }

}

exports.SigaaBondFactory = SigaaBondFactory;
