// sigaa-api/dist/search/sigaa-search-teacher-result.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaSearchTeacherResult = void 0;

require("source-map-support/register");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * @category Internal
 */
class SigaaSearchTeacherResult {
  constructor(http, parser, options) {
    this.http = http;
    this.parser = parser;

    _defineProperty(this, "_name", void 0);

    _defineProperty(this, "_department", void 0);

    _defineProperty(this, "_pageURL", void 0);

    _defineProperty(this, "_photoURL", void 0);

    this._name = options.name;
    this._department = options.department;
    this._pageURL = options.pageURL;
    this._photoURL = options.photoURL;
  }

  async getEmail() {
    const page = await this.http.get(this.pageURL.href);
    const contactElements = page.$('#contato').children().toArray();
    let email;

    for (const contactElement of contactElements) {
      const name = this.parser.removeTagsHtml(page.$(contactElement).find('dt').html());

      if (name === 'Endereço eletrônico') {
        email = this.parser.removeTagsHtml(page.$(contactElement).find('dd').html());
        break;
      }
    }

    if (email && email !== 'não informado') {
      return email;
    } else {
      return undefined;
    }
  }

  get name() {
    return this._name;
  }

  get profilePictureURL() {
    return this._photoURL;
  }

  downloadProfilePicture(basepath, callback) {
    if (!this.profilePictureURL) throw new Error("SIGAA: This teacher doesn't have profile picture");
    return this.http.downloadFileByGet(this.profilePictureURL.href, basepath, callback);
  }

  get department() {
    return this._department;
  }

  get pageURL() {
    return this._pageURL;
  }

}

exports.SigaaSearchTeacherResult = SigaaSearchTeacherResult;
