// sigaa-api/dist/account/sigaa-account-factory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaAccountFactory = void 0;

require("source-map-support/register");

var _sigaaAccountIfsc = require("./sigaa-account-ifsc");

var _sigaaAccountUfpb = require("./sigaa-account-ufpb");

var _sigaaAccountUnb = require("./sigaa-account-unb");

var _sigaaAccountUffs = require("./sigaa-account-uffs");

/**
 *
 * Serves to create account instances.
 * @category Internal
 */
class SigaaAccountFactory {
  constructor(http, parser, session, bondFactory) {
    this.http = http;
    this.parser = parser;
    this.session = session;
    this.bondFactory = bondFactory;
  }
  /**
   * Creates a new instance of Account.
   * @param page home page of account (page after login).
   */


  async getAccount(page) {
    const SigaaAccountInstitution = {
      IFSC: _sigaaAccountIfsc.SigaaAccountIFSC,
      UFPB: _sigaaAccountUfpb.SigaaAccountUFPB,
      UNB: _sigaaAccountUnb.SigaaAccountUNB,
      UFFS: _sigaaAccountUffs.SigaaAccountUFFS
    };
    return new SigaaAccountInstitution[this.session.institution](page, this.http, this.parser, this.session, this.bondFactory);
  }

}

exports.SigaaAccountFactory = SigaaAccountFactory;
