// sigaa-api/dist/courses/resources/forum/sigaa-course-forum-student.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SigaaCourseForum = void 0;

require("source-map-support/register");

var _url = require("url");

var _formdataNode = require("formdata-node");

var _updatableResource = require("../../../resources/updatable-resource");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * @category Internal
 */
class SigaaCourseForum extends _updatableResource.AbstractUpdatableResource {
  /**
   * Type of instance
   */

  /**
   * Form with parameters and url to load the forum page.
   */

  /**
   * Title of the forum.
   */

  /**
   * Number of topics
   */

  /**
   * Forum creator.
   */

  /**
   * Like 'Uma única discussão simples'
   */

  /**
   * Description of the forum.
   */

  /**
   * Form to submit topics in the forum.
   */

  /**
   * File attached to the forum.
   */

  /**
   * the forum id.
   */

  /**
   * If read monitoring is enabled.
   */

  /**
   *
   */

  /**
   * creation date of the forum.
   */
  constructor(http, parser, courseResourcesFactory, forumOptions, updater) {
    super(forumOptions.instanceIndentifier, updater);
    this.http = http;
    this.parser = parser;
    this.courseResourcesFactory = courseResourcesFactory;

    _defineProperty(this, "type", 'forum');

    _defineProperty(this, "_form", void 0);

    _defineProperty(this, "_title", void 0);

    _defineProperty(this, "_numOfTopics", void 0);

    _defineProperty(this, "_author", void 0);

    _defineProperty(this, "_forumType", void 0);

    _defineProperty(this, "_description", void 0);

    _defineProperty(this, "_submitTopicPageForm", void 0);

    _defineProperty(this, "_file", void 0);

    _defineProperty(this, "_id", void 0);

    _defineProperty(this, "_flagMonitorReading", void 0);

    _defineProperty(this, "_fullForumPromise", null);

    _defineProperty(this, "_creationDate", void 0);

    this.update(forumOptions);
  }

  update(forumOptions) {
    this._title = forumOptions.title;
    this._form = forumOptions.form;
    this._id = forumOptions.id;

    if (forumOptions.forumType !== undefined) {
      this._forumType = forumOptions.forumType;
    }

    if (forumOptions.numOfTopics !== undefined) {
      this._numOfTopics = forumOptions.numOfTopics;
    }

    if (forumOptions.author !== undefined) {
      this._author = forumOptions.author;
    }
  }
  /**
   * Loads the forum page.
   * @param retry
   */


  async getForumPage(retry = true) {
    try {
      const page = await this.http.post(this._form.action.href, this._form.postValues);
      this.parseForumTable(page);
      this.parseSubmitPageForm(page);
    } catch (err) {
      if (retry) {
        await this.updateInstance();
        return this.getForumPage(false);
      } else {
        throw err;
      }
    }
  }
  /**
   * Parse submit form in forum page.
   * @param page
   */


  parseSubmitPageForm(page) {
    const formElement = page.$('form#form');
    const action = formElement.attr('action');
    if (!action) throw new Error('SIGAA: Forum submit page has form without action.');
    const actionURL = new _url.URL(action, page.url.href);
    const postValues = {};
    formElement.find("input:not([type='button'])").each((index, element) => {
      const name = page.$(element).attr('name');
      if (name) postValues[name] = page.$(element).val();
    });
    this._submitTopicPageForm = {
      action: actionURL,
      postValues
    };
  }
  /**
   * Parse main page of forum.
   * @param page
   */


  parseForumTable(page) {
    const tableElement = page.$('table.formAva > tbody');
    if (tableElement.length === 0) throw new Error('SIGAA: Unexpected forum page without table element.');
    const rows = tableElement.find('tr').toArray();

    for (const row of rows) {
      const headCellElement = page.$(row).find('th');
      const dataCellElement = page.$(row).find('td');
      const label = this.parser.removeTagsHtml(headCellElement.html());
      const content = this.parser.removeTagsHtml(dataCellElement.html());

      switch (label) {
        case 'Título:':
          {
            this._title = content;
            break;
          }

        case 'Descrição:':
          {
            this._description = content;
            break;
          }

        case 'Autor(a):':
          {
            this._author = content;
            break;
          }

        case 'Arquivo:':
          {
            const linkElement = page.$(dataCellElement).find('a');

            if (linkElement.length === 1) {
              const title = this.parser.removeTagsHtml(linkElement.html());
              const onClick = linkElement.attr('onclick');
              if (!onClick) throw new Error('SIGAA: Invalid file format at forum page.');
              const form = page.parseJSFCLJS(onClick);
              const fileObj = {
                title,
                description: '',
                form,
                id: form.postValues.id,
                instanceIndentifier: form.postValues.id
              };

              if (this._file) {
                this._file.update(fileObj);
              } else {
                this._file = this.courseResourcesFactory.createFileFromFileData(fileObj, this.http, async () => {
                  await this.getForumPage();
                });
              }
            } else {
              this._file = undefined;
            }

            break;
          }

        case 'Monitorar Leitura:':
          {
            if (content === 'SIM') {
              this._flagMonitorReading = true;
            } else {
              this._flagMonitorReading = false;
            }

            break;
          }

        case 'Tipo:':
          {
            this._forumType = content;
            break;
          }

        case 'Ordenação Padrão:':
          {
            //TODO
            break;
          }

        case 'Criado em:':
          {
            const dates = this.parser.parseDates(content, 1);
            this._creationDate = dates[0];
            break;
          }

        default:
          {
            console.log('WARNING:forum label not recognized:' + label);
          }
      }
    }
  }

  async loadForumPage() {
    if (!this._fullForumPromise) {
      this._fullForumPromise = this.getForumPage();

      this._fullForumPromise.finally(() => {
        this._fullForumPromise = null;
      });
    }

    return this._fullForumPromise;
  }

  get title() {
    this.checkIfItWasClosed();
    return this._title;
  }
  /**
   * Like 'Uma única discussão simples'
   */


  async getForumType() {
    this.checkIfItWasClosed();

    if (this._forumType === undefined) {
      await this.loadForumPage();
    }

    if (!this._forumType) throw new Error('SIGAA: Forum type could not be loaded.');
    return this._forumType;
  }

  async getDescription() {
    this.checkIfItWasClosed();

    if (this._description === undefined) {
      await this.loadForumPage();
    }

    if (!this._description) throw new Error('SIGAA: Forum description could not be loaded.');
    return this._description;
  }
  /**
   * Post author
   */


  async getAuthor() {
    this.checkIfItWasClosed();

    if (this._author === undefined) {
      await this.loadForumPage();
    }

    if (!this._author) throw new Error('SIGAA: Forum author could not be loaded.');
    return this._author;
  }
  /**
   * If notifications are enabled
   */


  async getFlagMonitorReading() {
    this.checkIfItWasClosed();

    if (this._flagMonitorReading === undefined) {
      await this.loadForumPage();
    }

    if (!this._flagMonitorReading) throw new Error('SIGAA: Forum monitor reading flag could not be loaded.');
    return this._flagMonitorReading;
  }
  /**
   * creation date of the forum.
   */


  async getCreationDate() {
    this.checkIfItWasClosed();

    if (this._creationDate === undefined) {
      await this.loadForumPage();
    }

    if (!this._creationDate) throw new Error('SIGAA: Forum creation date could not be loaded.');
    return this._creationDate;
  }
  /**
   * File attached to the forum.
   */


  async getFile() {
    this.checkIfItWasClosed();

    if (this._file === undefined) {
      await this.loadForumPage();
    }

    return this._file;
  }
  /**
   * Number of topics
   */


  async getNumOfTopics() {
    this.checkIfItWasClosed();

    if (this._numOfTopics === undefined) {
      await this.updateInstance();
    }

    if (!this._numOfTopics) throw new Error('SIGAA: Forum number of topics could not be loaded.');
    return this._numOfTopics;
  }

  get id() {
    this.checkIfItWasClosed();
    return this._id;
  }
  /**
   * Post topic in forum
   * @param title title of topic
   * @param body body of topic
   * @param file buffer of file attachment
   * @param notify if notify members
   */


  async postTopic(title, body, file, notify) {
    if (!title) {
      throw new Error('SIGAA: Title topic forum cannot be empty.');
    }

    if (!body) {
      throw new Error('SIGAA: Title body forum cannot be empty.');
    }

    if (!this._submitTopicPageForm) {
      await this.loadForumPage();
    }

    if (!this._submitTopicPageForm) throw new Error('SIGAA: Could not get the forum form.');
    const page = await this.http.post(this._submitTopicPageForm.action.href, this._submitTopicPageForm.postValues);
    const formElement = page.$('form#form');
    const action = formElement.attr('action');
    if (!action) throw new Error('SIGAA: Forum post page has form without action.');
    const actionURl = new _url.URL(action, page.url.href);
    const inputHiddens = formElement.find('form#form input[type="hidden"]').toArray();
    const fileInput = formElement.find('input[type="file"]');
    const submitButton = formElement.find('input[name="form:btnSalvar"]');
    const notifyCheckbox = formElement.find('input[type="checkbox"]');
    if (inputHiddens.length === 0 || submitButton.length !== 1 || notifyCheckbox.length !== 1 || fileInput.length !== 1) throw new Error('SIGAA: Forum post page format is different than expected.');
    const formData = new _formdataNode.FormData();

    for (const input of inputHiddens) {
      const name = page.$(input).attr('name');
      if (name) formData.set(name, page.$(input).val());
    }

    if (file) {
      const name = fileInput.attr('name');
      if (!name) throw new Error('SIGAA: Forum post page has input file without name.');
      formData.set(name, file);
    }

    if (notify) {
      const name = notifyCheckbox.attr('name');
      if (!name) throw new Error('SIGAA: Forum post page has notify checkbox without name.');
      formData.set(name, 'on');
    }

    formData.set('form:assunto', title);
    formData.set('form:mensagem', body);
    const sumbitName = page.$(submitButton).attr('name');
    if (!sumbitName) throw new Error('SIGAA: Forum post page has submit button without name.');
    formData.set(sumbitName, page.$(submitButton).val());
    const responsePage = await this.http.postMultipart(actionURl.href, formData);

    if (!responsePage.bodyDecoded.includes('Operação realizada com sucesso!')) {
      throw new Error('SIGAA: Unexpected response forum post page.');
    }
  }

  *getTopics() {
    let stillHaveTopics = true;

    while (stillHaveTopics) {
      yield (async () => {
        stillHaveTopics = false;
        return true;
      })();
    }
  }

}

exports.SigaaCourseForum = SigaaCourseForum;
